packages/core/src/application.ts

```diff
@@ -293,125 +50,118 @@ export class Application extends Context {
-      path,
-      spec,
-      controller,
-      methodName));
+  public servers<T extends Server>(ctors: Constructor<T>[]) {
+    ctors.map(ctor => this.server(ctor));
   }
 
-  api(spec: OpenApiSpec): Binding {
-    return this.bind(CoreBindings.API_SPEC).to(spec);
+  /**
+   * Retrieve the singleton instance for a bound constructor.
+   *
+   * @template T
+   * @param {Constructor<T>=} ctor The constructor that was used to make the
+   * binding.
+   * @returns {Promise<T>}
+   * @memberof Application
+   */
+  public async getServer<T extends Server>(
+    target: Constructor<T> | String,
+  ): Promise<T> {
+    const key =
+      target instanceof String
+        ? `components.${target}.server`
+        : `components.${target.name}.server`;
+    return (await this.get(key)) as T;
   }
 
   /**
-   * Get the OpenAPI specification describing the REST API provided by
-   * this application.
+   * Retrieve the singleton instance for a bound constructor synchronously.
+   * Whenever possible, you should make use of the async "getServer" instead.
    *
-   * This method merges operations (HTTP endpoints) from the following sources:
-   *  - `app.api(spec)`
-   *  - `app.controller(MyController)`
-   *  - `app.route(route)`
-   *  - `app.route('get', '/greet', operationSpec, MyController, 'greet')`
+   * @template T
+   * @param {Constructor<T>} ctor
+   * @returns {T}
+   * @memberof Application
    */
-  getApiSpec(): OpenApiSpec {
-    const spec = this.getSync(CoreBindings.API_SPEC);
-
-    // Apply deep clone to prevent getApiSpec() callers from
-    // accidentally modifying our internal routing data
-    spec.paths = cloneDeep(this.httpHandler.describeApiPaths());
+  public getServerSync<T extends Server>(ctor: Constructor<T>): T {
+    return this.getSync(`components.${ctor.name}.server`) as T;
+  }
 
-    return spec;
+  /**
+   * Start the application, and all of its registered servers.
+   *
+   * @returns {Promise}
+   * @memberof Application
+   */
+  public async start(): Promise<void> {
+    const bindings = this.find('components.*.server');
+    return Promise.all(
+      bindings.map(async binding => {
+        const server = binding.getValue(this) as Server;
+        return server.start();
+      }),
+    ).then(() => {
+      return Promise.resolve();
+    });
   }
 
-  protected _logError(
-    err: Error,
-    statusCode: number,
-    req: ServerRequest,
-  ): void {
-    console.error(
-      'Unhandled error in %s %s: %s %s',
-      req.method,
-      req.url,
-      statusCode,
-      err.stack || err,
-    );
+  /**
+   * Stop the application instance and all of its registered servers.
+   * @returns {Promise}
+   * @memberof Application
+   */
+  public async stop(): Promise<void> {
+    const bindings = this.find('components.*.server');
+    return Promise.all(
+      bindings.map(async binding => {
+        const server = binding.getValue(this) as Server;
+        return server.stop();
+      }),
+    ).then(() => {
```

Comment:
```ts
await Promise.all(
  // ...
);
```

I find this code difficult to read because it's mixing several steps together: iterate all server bindings, instantiate a server, start a server. I am proposing to extract a shared helper function, e.g.

```ts
public async stop(): Promise<void> {
  await this._forEachServer(s => await s.stop());
}

protected _forEachServer<T>(fn: (s: Server) => Promise<T>) {
  const bindings = this.find('components.*.server');
  return await Promise.all(bindings.map(async binding => {
    const server = await Promise.resolve(binding.getValue(this)) as Server;
    return await fn(server);
  });
}
```

Note that `binding.getValue` can return either the resolved value or a promise. `ctx.get` handles these two cases for you. If you want to call `binding.getValue`  then you need to take care of it yourself. I personally think we should use `this.get(binding.key)`.